File: src/interception/simulation.py
import math
import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional

from .constants import DEFAULT_HEADLESS_MAX_STEPS, SIM_DT
from .params import ScenarioParams
from .state import State
from .telemetry import TelemetryRecorder

SEED_SPACE = 2**32


class EpisodeEnd(str, Enum):
    """Why the episode stopped."""

    NONE = "none"
    SUCCESS = "success"
    TIMEOUT = "timeout"


@dataclass
class SimulationConfig:
    """Everything needed to reproduce an episode exactly."""

    dt: float = SIM_DT
    max_steps: int = 0  # 0 = no cap (interactive); >0 enforces TIMEOUT
    seed: Optional[int] = None  # None = draw one and record it
    scenario: ScenarioParams = field(default_factory=ScenarioParams)
    record_telemetry: bool = False


class Simulation:
    """
    Owns world state and the episode lifecycle: reset -> step until terminal.

    Core logic is pygame-free; rendering reads ``state`` from the outside.
    """

    def __init__(self, config: Optional[SimulationConfig] = None):
        self.config = config or SimulationConfig()
        self.dt = self.config.dt
        self.max_steps = self.config.max_steps
        # Resolved once and stored, so an unseeded run is still reportable
        # and replayable after the fact.
        self.seed = self._resolve_seed(self.config.seed)
        self.state = State(seed=self.seed, scenario=self.config.scenario)
        self.telemetry = TelemetryRecorder() if self.config.record_telemetry else None
        self.steps = 0
        self.done = False
        self.end_reason = EpisodeEnd.NONE

    @staticmethod
    def _resolve_seed(seed):
        if seed is not None:
            return int(seed)
        return random.SystemRandom().randrange(SEED_SPACE)

    @property
    def elapsed_s(self):
        return self.steps * self.dt

    @property
    def guidance_law(self):
        return self.config.scenario.guidance.law

    def reset(self, seed: Optional[int] = None) -> None:
        """Rebuild the world. ``seed=None`` replays the current seed."""
        if seed is not None:
            self.seed = int(seed)
        self.state.reset(self.seed)
        if self.telemetry is not None:
            self.telemetry = TelemetryRecorder()
        self.steps = 0
        self.done = False
        self.end_reason = EpisodeEnd.NONE

    def step(self) -> None:
        """Advance one fixed timestep of ``dt`` seconds."""
        if self.done:
            return
        self.state.update(self.dt)
        self.steps += 1
        if self.telemetry is not None:
            self.telemetry.capture(self)
        if not self.state.targets:
            self.done = True
            self.end_reason = EpisodeEnd.SUCCESS
        elif self.max_steps > 0 and self.steps >= self.max_steps:
            self.done = True
            self.end_reason = EpisodeEnd.TIMEOUT

    def run(self):
        """Step until terminal. Requires ``max_steps > 0`` to be bounded."""
        while not self.done:
            self.step()
        return self

    def observation(self) -> dict:
        """Flat, JSON-friendly episode summary for logging and Monte Carlo."""
        miss = self.state.min_miss_distance_m
        return {
            "seed": self.seed,
            "guidance": self.guidance_law,
            "step": self.steps,
            "elapsed_s": round(self.elapsed_s, 4),
            "done": self.done,
            "end_reason": self.end_reason.value,
            "intercepts": self.state.intercepts,
            "min_miss_distance_m": None if math.isinf(miss) else round(miss, 4),
            "delta_v_mps": round(sum(a.delta_v_mps for a in self.state.agents), 3),
            "num_targets": len(self.state.targets),
            "num_agents": len(self.state.agents),
            "num_obstacles": len(self.state.obstacles),
        }


def run_headless(
    seed: Optional[int] = None,
    max_steps: int = DEFAULT_HEADLESS_MAX_STEPS,
    scenario: Optional[ScenarioParams] = None,
    record_telemetry: bool = False,
) -> Simulation:
    """Run until SUCCESS or TIMEOUT; returns the simulation for inspection."""
    cfg = SimulationConfig(
        seed=seed,
        max_steps=max_steps,
        scenario=scenario or ScenarioParams(),
        record_telemetry=record_telemetry,
    )
    return Simulation(cfg).run()

File: core/src/simulation.cpp
#include "interception/simulation.hpp"

#include <ostream>
#include <random>
#include <stdexcept>
#include <utility>

#include "interception/telemetry.hpp"

namespace interception {

namespace {

/// Draw an arbitrary seed. Deliberately outside the reproducible stream --
/// it only *chooses* which stream to replay, and the choice is recorded.
std::uint64_t drawSeed() {
    std::random_device device;
    return (static_cast<std::uint64_t>(device()) << 32) ^ device();
}

}  // namespace

std::string episodeEndName(EpisodeEnd end) {
    switch (end) {
        case EpisodeEnd::None:
            return "none";
        case EpisodeEnd::Success:
            return "success";
        case EpisodeEnd::Timeout:
            return "timeout";
    }
    throw std::invalid_argument("unhandled episode end");
}

std::ostream& operator<<(std::ostream& os, EpisodeEnd end) {
    return os << episodeEndName(end);
}

Simulation::Simulation(SimulationConfig config)
    : config_(std::move(config)),
      seed_(config_.hasSeed ? config_.seed : drawSeed()),
      state_(std::make_unique<State>(seed_, config_.scenario)) {
    if (config_.recordTelemetry) {
        telemetry_ = std::make_unique<TelemetryRecorder>();
    }
}

Simulation::~Simulation() = default;
Simulation::Simulation(Simulation&&) noexcept = default;
Simulation& Simulation::operator=(Simulation&&) noexcept = default;

void Simulation::reset() { reset(seed_); }

void Simulation::reset(std::uint64_t seed) {
    seed_ = seed;
    state_->reset(seed_);
    if (telemetry_) {
        telemetry_ = std::make_unique<TelemetryRecorder>();
    }
    steps_ = 0;
    done_ = false;
    endReason_ = EpisodeEnd::None;
}

void Simulation::step() {
    if (done_) {
        return;
    }
    state_->update(config_.dt);
    ++steps_;
    if (telemetry_) {
        telemetry_->capture(*this);
    }
    if (state_->targets().empty()) {
        done_ = true;
        endReason_ = EpisodeEnd::Success;
    } else if (config_.maxSteps > 0 && steps_ >= config_.maxSteps) {
        done_ = true;
        endReason_ = EpisodeEnd::Timeout;
    }
}

Simulation& Simulation::run() {
    while (!done_) {
        step();
    }
    return *this;
}

Observation Simulation::observation() const {
    Observation obs;
    obs.seed = seed_;
    obs.guidance = guidanceLawName(config_.scenario.guidance.law);
    obs.step = steps_;
    obs.elapsedS = elapsedS();
    obs.done = done_;
    obs.endReason = episodeEndName(endReason_);
    obs.intercepts = state_->intercepts();
    obs.hasMinMissDistance = state_->hasMinMissDistance();
    obs.minMissDistanceM = obs.hasMinMissDistance ? state_->minMissDistanceM() : 0.0;
    obs.deltaVMps = state_->totalDeltaVMps();
    obs.numTargets = static_cast<int>(state_->targets().size());
    obs.numAgents = static_cast<int>(state_->agents().size());
    obs.numObstacles = static_cast<int>(state_->obstacles().size());
    return obs;
}

Simulation runHeadless(std::uint64_t seed, int maxSteps, ScenarioParams scenario,
                       bool recordTelemetry) {
    SimulationConfig config;
    config.hasSeed = true;
    config.seed = seed;
    config.maxSteps = maxSteps;
    config.scenario = std::move(scenario);
    config.recordTelemetry = recordTelemetry;

    Simulation sim(std::move(config));
    sim.run();
    return sim;
}

}  // namespace interception
